core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalReadConverterContext.scala

```diff
@@ -37,39 +38,106 @@ class InternalReadConverterContext(override val keyName: String, override val ch
+
+  private def categoricalString(chunk: Chunk) = chunk.vec().domain()(chunk.at8(rowIdx).toInt)
+
+  private def uuidString(chunk: Chunk) = new UUID(chunk.at16h(rowIdx), chunk.at16l(rowIdx)).toString
+
+  private def plainString(chunk: Chunk) = {
+    chunk.atStr(new BufferedString(), rowIdx).toString
+  }
+
+  val StringProviders = Map[Byte, (Chunk => String)](
+    Vec.T_CAT -> categoricalString,
+    Vec.T_UUID -> uuidString,
+    Vec.T_STR -> plainString
+  ) withDefault((t:Byte) => {
+    assert(assertion = false, s"Should never be here, type is $t")
+    // TODO(vlad): this is temporarily here, to provide None is string is missing
+    (_:Chunk) => null
+  }
+  )
+
+  private def stringProvider(columnNum: Int): (Chunk => String) = StringProviders(fr.vec(columnNum).get_type())
+
+  override def getString(columnNum: Int): Option[String] = getOption(stringProvider(columnNum))(columnNum)
 
   private def underlyingFrame = DKV.get(Key.make(keyName)).get.asInstanceOf[Frame]
 
   override def numRows: Int = nrows
+
+  private val DefaultReader: Reader = Reader("DEFAULT", (_:Int) => None)
+
+  /**
+    * Given a a column number, returns an Option[T]
+    * with the value parsed according to NameOfType.
+    * You can override it.
+    *
+    * A map from type name to option reader
+    */
+  lazy val OptionReaders: Map[NameOfType, Int => Option[Any]] = Map(
+    "Boolean"    -> getBoolean,
+    "Byte"       -> getByte,
+    "Double"     -> getDouble,
+    "Float"      -> getFloat,
+    "Integer"    -> getInt,
+    "Long"       -> getLong,
+    "Short"      -> getShort,
+    "String"     -> getString,
+    "UTF8String" -> getUTF8String,
+    "Timestamp"  -> getTimestamp)
+
+  /**
+    * Given a type name, returns a default value for the case it's NaN in spark
+    * You can override it.
+    */
+  val Defaults: Map[NameOfType, Any] = Map(
+    "Boolean"   -> false,
+    "Byte"      -> 0.toByte,
+    "Double"    -> Double.NaN,
+    "Float"     -> Float.NaN,
+    "Integer"   -> 0,
+    "Long"      -> 0L,
+    "Short"     -> 0.toShort,
+    "String"    -> null,
+    "UTFString" -> null,
+    "Timestamp" -> 0L
+  )
+
+  private def twoReaders = (key: NameOfType, op: Int => Option[Any]) =>
+    List((s"Option[$key]":NameOfType) -> Reader(s"Option[$key]":NameOfType, op),
+      key                            -> Reader(key, (col: Int) => {
+        op.apply(col).getOrElse(Defaults(key))
+      })
+    )
+
+  private lazy val availableReadersByName: Map[NameOfType, Reader] =
+    OptionReaders flatMap twoReaders.tupled
+
+  lazy val readerMapByName: Map[NameOfType, Reader] = availableReadersByName withDefaultValue DefaultReader
+
+//  lazy val readerMap: Map[SupportedType, Reader] = availableReaders withDefaultValue DefaultReader
+
+  def columnValueProviders(columnIndexesWithTypes: Array[(Int, DataType)]): Array[() => Option[Any]] = {
+    for {
+      (i, ft) <- columnIndexesWithTypes
```

Comment:
`dataType` or something like that would be maybe more readable instead of `ft`. I know it's implied by the context, but still from my point of view things like this should be obvious after a first quick look :)